lib/krane/kubernetes_resource.rb

```diff
@@ -582,7 +581,7 @@ def validate_with_local_dry_run(kubectl)
       verb = deploy_method == :apply ? "apply" : "create"
       command = [verb, "-f", file_path, "--dry-run", "--output=name"]
       kubectl.run(*command, log_failure: false, output_is_sensitive: sensitive_template_content?,
-        retry_whitelist: [:client_timeout], attempts: 3, use_namespace: !global?)
```

Comment:
IIRC the reason for needing these whitelists was that the dry runs return a non-zero exit status when the validation ran successfully and found an invalid template. We do not want to re-run it in those cases. Before removing the whitelist, we need to prove I'm wrong about that in all supported versions of kubectl (and add a test showing it).

Reply:
@KnVerey What sort of invalid template?

Reply:
Any kind? We're doing this dry run to catch when the template is malformed, and when such an error is caught, the exit status is (or at least was when we built the feature) non-zero. Dry run success means no errors found in the templates.

Reply:
Honestly, I'm not sure how I would write that test. As a unit test on `KubernetesResource` digging into the guts of `Kubectl` to assert that it only shells out once?

I guess what you want is to separate the cases of non-zero with useful stderr (don't retry) from non-zero with empty stderr (retry). Would you be happy with a flag to `Kubectl#run` that we assert is passed and test it's behaviour in the `Kubectl` tests, or do you need a functional test for that, I guess using `stub_kubectl_response` in place of [mocking `Kubectl#run` as we do right now](https://github.com/Shopify/krane/blob/master/test/unit/krane/kubernetes_resource_test.rb#L272)?

Reply:
The first thing to do is to confirm what kubectl's actual behavior is, which you could do by dry-running a template with an error in it and noting the exit code. If it is indeed non-zero, then this PR needs to find another way to fix. As for the test, one idea would be to enable debug-level logging on a simple integration test with a single invalid resource. With debug-level logging enabled, [we print every command we run](https://github.com/Shopify/krane/blob/master/lib/krane/kubectl.rb#L35), so you could scan the output to see what actually happened.

Reply:
```
➭ cat foo.yaml
apiVersion: v1
metadata:
  name: foo
{19:25}~/Downloads [2.7.0]
➭ kubectl apply -f foo.yaml --dry-run
error: error validating "foo.yaml": error validating data: kind not set; if you choose to ignore these errors, turn validation off with --validate=false
{19:25}~/Downloads [2.7.0]
1 ↵ ➭ echo $?
1
{19:25}~/Downloads [2.7.0]
➭ kubectl apply -f foo.yaml --server-dry-run
error: error validating "foo.yaml": error validating data: kind not set; if you choose to ignore these errors, turn validation off with --validate=false
{19:26}~/Downloads [2.7.0]
1 ↵ ➭ echo $?
1
```

> If it is indeed non-zero, then this PR needs to find another way to fix.

I think maybe you misunderstood what I'm trying to fix here, @KnVerey . Sometimes I can dry run a totally valid resource, get a non-zero exit code, and empty stderr/stdout; a bogus failure. When this happens, I want the dry run to be retried; I'm using this change on a real-world project and it has completely solved our deployment failures. The errors only happen sporadically, and are bogus, and we shouldn't fail a deployment on them.

What about if I just retry the local-dry-run fallback, so as not to repeat connections to the API server for failed server-dry-run. Would that satisfy your requirement?

Reply:
I understand what you're trying to accomplish, but the result of the change on this PR is retrying _all_ validation failures, both legitimate and bogus, up to 3 times. In fact it guarantees that if you have an invalid template, the validation will be run 3 times. We have deploys for which local dry run is already noticeably slow as it is, and this change would degrade performance unacceptably. Looking at your bug report again, we could create a new [error matcher](https://github.com/Shopify/krane/blob/master/lib/krane/kubectl.rb#L6) that requires the string to be entirely blank, and add that to the retry whitelist.

Reply:
@KnVerey Does that last change look ok?

Reply:
Yep, that's what I had in mind